KIP-222 - Add Consumer Group operations to Admin API

jeqo opened the pull request:
KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-222+-+Add+Consumer+Group+operations+to+Admin+API

jeqo committed:
draft describeConsumerGroups on KafkaAdminClient

jeqo committed:
document changes

jeqo committed:
add `listConsumerGroups`

jeqo committed:
consumer group and instance class documentation

jeqo committed:
renaming to support groups (including connect groups)

jeqo committed:
merge

jeqo committed:
add list groups/consumer groups and list group offsets

jeqo committed:
increase class fan out, add test method unsupported exceptions

jeqo committed:
refactor listGroupOffsets

jeqo committed:
add support for offsetandmetadata as part of listGroupOffsets

jeqo committed:
refactor member description assignment to support extensibility

jeqo committed:
refactor member description assignment to support extensibility

jeqo committed:
fix checkstyle

jeqo committed:
improve documentation

jeqo committed:
fix bug: self assignment variable. typo

hachikuji commented:
I did not take a deep look yet, but there are some problems that will probably require changes to the KIP.  The group management protocol handles more than just consumer groups. Connect also uses it for its own group management. Connect groups do not have topic partition assignments, so the current `MemberAssignment` object does not work generally. I see two options:

1. Get rid of the generic group APIs and only support consumer groups for now. In other words, if the protocol type is not "consumer," we just raise an exception. This would be a bummer, but it's probably the only way to make your current API work.

2. Change the API to accommodate different group types. This will mean making the `MemberAssignment` generic so that it does not depend on the consumer assignment. 

Additionally, I think there are a few other points to consider:

- The describe API has the ability to fetch the assignment currently, but not the consumer's subscription. I think we should consider exposing this.
- Users typically expect "simple" consumer groups which do not use group management to be included when listing consumer groups. In this case, the protocol type will be empty.
- No fault of your's, but concurrently, there was a KIP to add support for deleting groups. We should consider adding an AdminClient API to support this (it could come later if needed, but there shouldn't be anything too tricky from an API perspective, so we may as well do it here.

hachikuji commented:
Just to clarify, if we go with the first option that I suggested above, we probably don't need a revote on the KIP. I doubt anyone will object if you just send an email to the discussion thread with the changes. Specifically, we should do the following:

1. Drop the `describeGroups` API.
2. Change the name of `listGroups` to `listConsumerGroups`.
3. Change `listGroupOffsets` to `listConsumerGroupOffsets`.

Otherwise, the logic stays the same. We would just need to filter groups which do not have the "consumer" protocol type. Note that I do think it's important to account for simple consumer groups as well. These can be identified as groups with an empty protocol type.

jeqo commented:
@hachikuji thanks for your feedback! 

Putting Group Management (connect and consumer groups) together as part of the changes to the API was one of the important points discussed in this KIP.

As I understand so far: connect group details are different from consumer group details, so we will need to define different APIs for both instead of trying to make them generic as your second option propose.

Following what your first option proposed we can add:

1. Drop `describeGroups` API (and change it to `describeConsumerGroups`?)
2. Change `listGroups` to `listConsumerGroups` 
3. Change `listGroupOffsets` to `listConsumerGroupOffsets`.
4. add `listConnectGroups`?
5. I don't know how Connect Groups use Group Management API in detail but if they have offset management also, we can add `listConnectGroupOffsets` and `describeConnectGroups` API

Simple Consumer Groups will be added to `listConsumerGroupOffsets` filtering protocol type = "consumer" or empty right? I can propose it as part of the KIP discussion to check if it make sense to add them as part of the same response.

And about adding DeleteGroups API I would be happy to, but I think we would need to hold until https://github.com/apache/kafka/pull/4479 is merged right?

hachikuji commented:
@jeqo I'd probably suggest sticking with consumer groups initially if we want to have any hope of getting this into the release. So if we have 1-3 in your list, we can leave general group APIs for future work. The nice thing about sticking with consumer groups is that it lets us expose the inner details from a consumer perspective. Here a couple concrete suggestions:

1. Change `GroupListing` to `ConsumerGroupListing`. If we do that, then we don't need to expose "protocol type" at all. Maybe we can replace it with a simple flag which indicates whether it is a simple group or not.
2. Change `GroupDescription` to `ConsumerGroupDescription`. We can do the same thing with the "protocol type." We should also expose the group "protocol," which for consumers, just means the assignment strategy.

As far as DeleteGroups, we should probably only consider it if we manage to get 1-3 merged.

Thoughts?

jeqo committed:
change scope of groups to consumergroups

jeqo committed:
fix removed method

jeqo commented:
@hachikuji I have applied changes mentioned above. Let me know WDYT, and if I should take this back to the KIP discussion to gather some feedback. I could propose to move Connect Group Management and Delete Consumer Group to a next KIP.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/ConsumerGroupDescription.java:
```diff
@@ -0,0 +1,104 @@
+    private final String groupId;
+    private final boolean isSimpleConsumerGroup;
+    private final List<MemberDescription> members;
+    private final String protocol;
```
Maybe we can use the consumer-specific term "partitionAssignor"?

jeqo replied:
:+1:

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1981,4 +1992,145 @@ void handleFailure(Throwable throwable) {
+        }
+        final long now = time.milliseconds();
+        runnable.call(new Call("describeConsumerGroups", calcDeadlineMs(now, options.timeoutMs()),
+            new ControllerNodeProvider()) {
```
The DescribeGroup API has to be sent to the group coordinator, which is potentially a different node for each group. You use the FindCoordinator API in order to lookup the coordinator for a given group. The logic should be something like this:

1. For each group in the request, send a FindCoordinator request to any node in the cluster.
2. Group the results by coordinator id.
3. Send DescribeGroups to each coordinator from 2.

Ideally, we should also handle retries correctly. It could happen that the coordinator moves to another node by the time we send DescribeGroups. In this case, the error code will be NOT_COORDINATOR. We should handle this by looking up the coordinator again.

jeqo replied:
OK, I have pushed draft following this description.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1981,4 +1992,145 @@ void handleFailure(Throwable throwable) {
+    }
+
+    @Override
+    public ListConsumerGroupsResult listConsumerGroups(ListConsumerGroupsOptions options) {
```
Unfortunately, the consumer groups are not aggregated in the same way that topic metadata is. To get all the groups in the cluster, you have to send the ListGroups request to all nodes.

jeqo replied:
I assume this is the same as `DescribeConsumerGroup` API. If changes are fine, I will updated this an `listConsumerGroupOffsets`

hachikuji replied:
The logic is different for ListGroups. We have to send a separate request to every broker in the cluster and then aggregate the results.

jeqo replied:
@hachikuji trying to test this I find out that the result from this operation has to change (similar to how the Scala API is doing now) to respond with `Map<Node, KafkaFuture<List<ConsumerGroupListing>>>` instead, to list all responses from nodes. Does this make sense? I can update the KIP regarding this.

guozhangwang replied:
I'd vote for doing the flattening internally than expose the node information in the results, as this is supposed to be internal implementation details that is better not leaking out. The Scala API returning `Map[Node, List[GroupOverview]]` was not a well designed one in my hind-sight.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1981,4 +1992,145 @@ void handleFailure(Throwable throwable) {
+    }
+
+    @Override
+    public ListConsumerGroupOffsetsResult listConsumerGroupOffsets(final String groupId, final ListConsumerGroupOffsetsOptions options) {
```
Like DescribeGroups, we need to find the coordinator for the group to send the OffsetFetch request to.

hachikuji reviewed (commented):
@jeqo Thanks for the updates. From an API perspective, this is looking good. However, the request logic needs a bit more work. I would suggest having a look at `kafka.admin.AdminClient` to understand the logic behind these requests.

I don't think it's reasonable to get this patch in the next day, so let's plan on getting this into the next feature release. With a little more time, we can also consider the delete API and non-consumer groups (potentially in a separate KIP).

jeqo committed:
rename protocol to partitionassignor

jeqo committed:
lookup for coordinator before describe consumer group

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/349/

jeqo committed:
Merge remote-tracking branch 'origin/trunk' into feature/admin-client-describe-consumer-group

jeqo committed:
lookup for coordinator before list consumer group and offset

hachikuji reviewed (commented):
Apologies for the review delay. I will look in more depth once the 1.1 release has been finalized. In the meantime, I noticed there were no test cases. It would be a good idea to add both unit and integration tests for the new functionality.

jeqo commented:
Yes, I will work on that during this and the next week :+1:

jeqo committed:
Merge remote-tracking branch 'origin/trunk' into feature/admin-client-describe-consumer-group

jeqo committed:
unit testing list consumer groups

jeqo committed:
missing unit tests

ijuma commented:
@jeqo Any updates?

jeqo committed:
Merge branch 'feature/admin-client-describe-consumer-group' of github.com:jeqo/kafka into feature/admin-client-describe-consumer-group

jeqo commented:
@ijuma I finally got some time to work on this these days. I will try to push a final version with unit and integration tests this week. Sorry for the delay

ijuma commented:
@jeqo Thanks!

jeqo committed:
Merge remote-tracking branch 'origin/trunk' into feature/admin-client-describe-consumer-group

jeqo committed:
unit testing kip-222

jeqo commented:
@hachikuji @ijuma I have add unit tests and add some changes to the APIs to wrap results in a `KafkaFuture` to handle first request to list nodes and find coordinators according to https://github.com/apache/kafka/pull/4454#discussion_r164891051 I will send a KIP update accordingly. Looking forward to your feedback

jeqo committed:
support delete consumer groups

jeqo committed:
fix test

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/ConsumerGroupListing.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public String toString() {
+        return "(isSimpleConsumerGroup=" + isSimpleConsumerGroup + ")";
```
Why not including groupId here?

jeqo replied:
You're right. I'll fix it.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupsResult.java:
```diff
@@ -0,0 +1,76 @@
+public class ListConsumerGroupsResult {
+    final KafkaFuture<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>> futureMap;
+
+    ListConsumerGroupsResult(KafkaFuture<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>> futureMap) {
```
Please see my other comment: I'd vote for doing the flattening internally than expose the node information in the results, as this is supposed to be internal implementation details that is better not leaking out. cc @hachikuji

jeqo replied:
Agree, but I was wondering: if a Node is not available, we could return a list with less consumer groups, right? In that case, a map with Node -> List[ConsumerGroupListing] could help to compare results against a list of Nodes.

If this is not an issue, I was thinking about changing the result to: `KafkaFuture<Collection<KafkaFuture<ConsumerGroupListing>>>` 
make sense?

guozhangwang replied:
That is a good point.

I think adding a boolean flag in addition to the `future` result indicating if it may be incomplete is better than exposing the node map. I.e. we could have two fields inside `ListConsumerGroupsResult`, a `boolean` and a `KafkaFuture<Collection<ConsumerGroupListing>>` (I think we do not need to have nested KafkaFutures so I did not have that inside `Collection`, but correct me if I overlook anything).

jeqo replied:
Agree. Would be OK to get list of nodes from `describeCluster` and then join futures?:

```
    private Collection<Node> listNodes(Integer timeoutMs) {
        try {
            return describeCluster(new DescribeClusterOptions().timeoutMs(timeoutMs)).nodes().get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }
```

```
        final Collection<Node> nodes = listNodes(options.timeoutMs);

        final Set<KafkaFutureImpl<Collection<ConsumerGroupListing>>> futures = new HashSet<>(nodes.size());

        for (final Node node : nodes) {
          ...
        }

        KafkaFuture.allOf(futures.toArray(new KafkaFuture[0]))
                .thenApply(new KafkaFuture.Function<Void, Collection<ConsumerGroupListing>>() {
                    @Override
                    public Collection<ConsumerGroupListing> apply(Void aVoid) {
                        final Set<ConsumerGroupListing> consumerGroups = new HashSet<>();
                        for (KafkaFutureImpl<Collection<ConsumerGroupListing>> future : futures) {
                            try {
                                consumerGroups.addAll(future.get());
                            } catch (InterruptedException | ExecutionException e) {
                                // This should be unreachable, because allOf ensured that all the futures
                                // completed successfully.
                                throw new RuntimeException(e);
                            }
                        }
                        return consumerGroups;
                    }
                });
```

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+            final long nowFindCoordinator = time.milliseconds();
+            final long deadline = calcDeadlineMs(nowFindCoordinator, options.timeoutMs());
+
+            runnable.call(new Call("findCoordinator", deadline, new ControllerNodeProvider()) {
```
Why we need to ask controller for the coordinator? Should we just ask any node? I.e. `LeastLoadedNodeProvider`. cc @cmccabe

jeqo replied:
agree.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+        final long nowFindCoordinator = time.milliseconds();
+        final long deadline = calcDeadlineMs(nowFindCoordinator, options.timeoutMs());
+
+        runnable.call(new Call("findCoordinator", deadline, new ControllerNodeProvider()) {
```
Ditto as above, we could use any node to find coordinator.

jeqo replied:
ok.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+                                }
+
+                                final String protocolType = groupMetadata.protocolType();
+                                if (protocolType.equals(ConsumerProtocol.PROTOCOL_TYPE) || protocolType.isEmpty()) {
```
This is a meta suggestion, not a comment for the changes of this PR: although currently in AK we only have two protocol types, it is possible that we will add more in AK (e.g. Kafka Streams may implement its own coordiantor), and also non-AK code can customize their own coordinator implementation with different protocol names. So it may be less appealing to add one API per each protocol type with such filtering in this line as it is 1) less extensible and 2) not usable for other non-builtin protocol types.

I'd suggest: 

a. In this PR we keep the `XXXConsumerGroups` APIs in admin client, and complete this as KIP-222.
b. In the future we can add another set of `XXXGroups(String protocolType, ..)` API.

The hope is that a) is commonly used for group queries and hence worth making it a special "syntax sugar" for most users, while only advanced users would like to use b) for other protocol types.

WDYT @jeqo @hachikuji ?

jeqo replied:
@guozhangwang agree. Actually I was wondering about `deleteConsumerGroups` as in the API there is no way (I'm aware of) to validate that we are deletign a *ConsumerGroup* and not another type of Group (e.g. Connect). Maybe this is a case to add `deleteGroups` instead of the current one?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+
+                @Override
+                void handleFailure(Throwable throwable) {
+                    if (throwable instanceof NotCoordinatorException) {
```
The find coordinator request should never see a `NotCoordinatorException` right?

jeqo replied:
agree.

jeqo committed:
fix toString ConsuerGroupListing

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+
+                    runnable.call(new Call("listConsumerGroups", deadline, new ConstantNodeIdProvider(brokerId)) {
+
+                        private final KafkaFutureImpl<Collection<ConsumerGroupListing>> future = entry.getValue();
```
I'd suggest flatten the map to abstract away which nodes contains which consumer groups as they are supposed to be internal information, we have the freedom to change those internal impl whenever we want. Once we expose such a public API it will be partially public information and hence hard to change.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+            final long nowFindCoordinator = time.milliseconds();
+            final long deadline = calcDeadlineMs(nowFindCoordinator, options.timeoutMs());
+
+            runnable.call(new Call("findCoordinator", deadline, new ControllerNodeProvider()) {
```
Ditto here.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+            }
+        }
+
+        for (final String groupId : groupIdList) {
```
Also I'm wondering if we could group the consumers after the first round of `findCoordinator` to send the delete group request once to each broker, than sending it once-at-a-time? Is this optimization worthwhile the complexity?

jeqo replied:
Not sure is worth it. As @cmccabe mentioned: 

> It's unfortunate that there's no batch version of findCoordinator -- it really makes batch operations less efficient than they would otherwise be. But that's not something we should try to fix in this PR-- we can do that later.

guozhangwang replied:
Makes sense.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+
+                        @Override
+                        AbstractRequest.Builder createRequest(int timeoutMs) {
+                            return new DeleteGroupsRequest.Builder(groupIdList);
```
Hmm.. is this correct that we send all the groups to each of the broker, i.e. many consumer groups may not be hosted by this broker right?

jeqo replied:
You're right. I would send a singleton `Set` for each `groupId`.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,318 @@ void handleFailure(Throwable throwable) {
+                                                ByteBuffer.wrap(Utils.readBytes(groupMember.memberAssignment())));
+
+                                        final MemberDescription memberDescription =
+                                            new MemberDescription(
```
This part is not correct: we are passing clientId as `consumerId`, and memberId as `clientId` to the MemberDescription.

We should just use the same string names, `clientId` and `memberId` in `MemberDescription` and pass in to constructors accordingly.

jeqo replied:
agree. fixing.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/MemberDescription.java:
```diff
@@ -0,0 +1,98 @@
+     * @param host       The host
+     * @param assignment The assignment
+     */
+    public MemberDescription(String consumerId, String clientId, String host, MemberAssignment assignment) {
```
See my other comment above: the first two parameters should be clientId and memberId.

jeqo replied:
fixing

guozhangwang commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -632,6 +642,190 @@ public void testDeleteRecords() throws Exception {
         }
     }
 
+    @Test
```
We do not have a unit test for delete group yet.

jeqo replied:
I will add one.

guozhangwang reviewed (commented):
Thanks @jeqo . Made a pass over the PR.

jeqo committed:
ask any node for coordinator node

jeqo committed:
fix memberDescription properties

jeqo committed:
fix to delete each group independently

cmccabe commented:
Thanks for this patch, [~jeqo].  The admin client parts look good overall.

It's unfortunate that there's no batch version of findCoordinator -- it really makes batch operations less efficient than they would otherwise be.  But that's not something we should try to fix in this PR-- we can do that later.

It seems like listConsumerGroupsOffsets should be a batch API, right?  It should be able to take multiple groupIds, not just a single one.  This would make it more consistent with the other APIs, and let us do batch optimizations later.

What is the purpose of `ConsumerGroupListing#isSimpleConsumerGroup`?  When would users need to check it?  Sorry if this question has been answered elsewhere.  However, I looked at the KIP and couldn't find anything about this.  We should probably update the KIP with this information.

We should be checking `PartitionData#error`, right?  It looks like we are not checking this currently.

guozhangwang commented:
> It seems like listConsumerGroupsOffsets should be a batch API, right? It should be able to take multiple groupIds, not just a single one. This would make it more consistent with the other APIs, and let us do batch optimizations later.

Yup, I agree with this one.



> What is the purpose of ConsumerGroupListing#isSimpleConsumerGroup? When would users need to check it? Sorry if this question has been answered elsewhere. However, I looked at the KIP and couldn't find anything about this. We should probably update the KIP with this information.

There's some discussion between @hachikuji and @jeqo . The conclusion is that in order to not expose the protocol type information we abstract it with this boolean



> We should be checking PartitionData#error, right? It looks like we are not checking this currently.

+1

cmccabe commented:
> > What is the purpose of ConsumerGroupListing#isSimpleConsumerGroup?

> There's some discussion between @hachikuji and @jeqo . The conclusion is that in order to not expose the protocol type information we abstract it with this boolean

Sorry if this is a dumb question, but can you give an example of when a user might want to check this boolean?

guozhangwang commented:
When the consumer is not participating in the membership management (i.e. groupID not set) on coordinator the flag will be true; in Streams for example, normal consumers rely on group membership management for partition assignment while restore consumer does not, but only rely on coordinator for offset management, i.e. is a "simple consumer group".

guozhangwang commented:
@jeqo is the current PR ready for review again?

jeqo committed:
remove inner Kafka Futures

jeqo commented:
@guozhangwang I have implement some changes: 

Remove KafkaFutures inside KafkaFuture. To do this a couple of private methods were added: `listNodes` and `findCoordinator`. 

About `ListConsumerGroupsResult` nodes, I have change the methods listings to return `KafkaFuture<Collection<ConsumerGroupListing>>`. WDYT?

Missing TODOs mentioned by @cmccabe I will work on:

- [x] We should be checking PartitionData#error, right? It looks like we are not checking this currently.
- [ ] It seems like listConsumerGroupsOffsets should be a batch API, right?

jeqo committed:
handle partition data errors

cmccabe commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2070,4 +2086,274 @@ void handleFailure(Throwable throwable) {
+
+    @Override
+    public ListConsumerGroupsResult listConsumerGroups(ListConsumerGroupsOptions options) {
+        final Collection<Node> nodes = listNodes(options.timeoutMs());
```
This is not correct.  It's blocking, which turns this into a blocking API rather than a non-blocking one.

guozhangwang replied:
+1

cmccabe commented:
> Remove KafkaFutures inside KafkaFuture. To do this a couple of private methods were added: listNodes and findCoordinator.

Please get rid of these functions.  They are blocking calls, and can't be called inside a non-blocking function.

guozhangwang commented:
> Remove KafkaFutures inside KafkaFuture. To do this a couple of private methods were added: listNodes and findCoordinator.

I'm not sure if I get the motivation of doing this completely, could you explain?

> About ListConsumerGroupsResult nodes, I have change the methods listings to return KafkaFuture<Collection<ConsumerGroupListing>>

That sounds good to me.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2100,223 +2141,164 @@ public DescribeConsumerGroupsResult describeConsumerGroups(final Collection<Stri
-                MetadataResponse metadataResponse = (MetadataResponse) abstractResponse;
-
-                final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futures = new HashMap<>();
+        final Map<Node, KafkaFutureImpl<Collection<ConsumerGroupListing>>> futures = new HashMap<>();
```
We can still follow the original pattern of calling a `listNodes` to any broker, get the result, and inside the response handling logic based on the result sending a `listConsumerGroups` to each node, and then group and flatten the result so we hide the node information of each sub-map it is belonging to.

guozhangwang reviewed (commented):
I'm with @cmccabe about the latest changes: we should not call `nodes().get()` since `Future#get()` is a blocking call.

jeqo committed:
revert remove inner Kafka Futures

jeqo commented:
@guozhangwang the idea was to remove the complexity of dealing with `KafkaFuture` wrapping a Map of `KafkaFuture` as values in the case of `listConsumerGroups`. 

Also, to flatten `Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>` to `KafkaFuture<Collection<ConsumerGroupListing>>`. 

If we have a `KafkaFuture<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>>` then `listings()` method will return: `KafkaFuture<KafkaFuture<Collection...>>`. Is this ok? Maybe I'm missing something about dealing with KafkaFutures here.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupsResult.java:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Return a future which yields a map of consumer groups to ConsumerGroupListing objects.
+     */
+    public KafkaFuture<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>> nodesToListings() {
```
As I suggested before, to not expose the node information, we should remove this function.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupsResult.java:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Return a future which yields a collection of ConsumerGroupListing objects.
+     */
+    public KafkaFuture<Collection<KafkaFuture<Collection<ConsumerGroupListing>>>> listings() {
```
I agree with you that we'd better return a `KafkaFuture<Collection<ConsumerGroupListing>>`; plus I think it is not only for one that would better be modified: e.g. `DescribeConsumerGroupsResult#all()` returns the same pattern as it also need to require asking to multiple consumers.

I spent some time into `KafkaFuture` a bit, and here is a general proposal I had for these two APIs:

1) we did the first blocking call internally after the first round trip returns immediately, note that although it is blocking an admin client thread, it is not blocking the user thread.

2) we flatten the futures into a `KafkaFuture<Collection<XX>>` with the `allOf...thenApply` in the first step.

Here is a PR I wrote to illustrate my idea (it is incomplete, only showing the idea based on ConsumerGroupListing): https://github.com/jeqo/kafka/pull/2

By doing that I'd suggest we just return `KafkaFuture<Collection<ConsumerGroupListing>>` for `ListConsumerGroupsResult.listings()`, and `KafkaFuture<Collection<ConsumerGroupDescription>>` in `DescribeConsumerGroupsResult#all()`.

cc @cmccabe

guozhangwang committed:
Merge branch 'feature/admin-client-describe-consumer-group' of https://github.com/jeqo/kafka into KMinor-improve-kafka-future

guozhangwang committed:
block and flatten after the first iteration

guozhangwang reviewed (commented):
> If we have a KafkaFuture<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>> then listings() method will return: KafkaFuture<KafkaFuture<Collection...>>. Is this ok? Maybe I'm missing something about dealing with KafkaFutures here.

I think I know your confusions now: I think we should modify the function of `listings()` to have two phases of blocking `get`. I'll leave a detailed comment at the code.

vvcephei committed:
MINOR: fix streams test-utils dependencies (#4821)

Reviewers: Bill Bejeck <bill@confluent.io>, Matthias J. Sax <matthias@confluent.io>, Guozhang Wang <wangguoz@gmail.com>

jeqo committed:
Merge remote-tracking branch 'origin/trunk' into feature/admin-client-describe-consumer-group

jeqo committed:
Merge branch 'KMinor-improve-kafka-future' of https://github.com/guozhangwang/kafka into guozhangwang-KMinor-improve-kafka-future

jeqo committed:
Merge branch 'guozhangwang-KMinor-improve-kafka-future' into feature/admin-client-describe-consumer-group

guozhangwang commented:
@jeqo I was pondering on my PR now and I realized it might not be correct, depending on whether we can call `get()` from the client thread directly, but the I'd still recommend you consider if this general idea works or not: after we have received the response from the listNodes request, we can let the client thread to flatten the map of futures into a single future and then send it back, i.e. we let the client thread to make one blocking call, so that the user thread, after getting the `result`, only need to make one more blocking call.

guozhangwang commented:
Note that the admin client uses a single thread which iteratively calls `poll` to send request and gets response, so for a `blocking` call we still need to let the thread to go back to its main loop while waiting for the results in order to flatten the futures.

chia7712 committed:
MINOR: Don’t send the DeleteTopicsRequest for invalid topic names (#4763)

The invalid topic name is already handled locally so it is unnecessary to send the DeleteTopicsRequest. This PR adds a count to MockClient for testing.

Reviewers: Colin Patrick McCabe <colin@cmccabe.xyz>, Jason Gustafson <jason@confluent.io>

guozhangwang commented:
> @jeqo I was pondering on my PR now and I realized it might not be correct, depending on whether we can call get() from the client thread directly, but the I'd still recommend you consider if this general idea works or not: after we have received the response from the listNodes request, we can let the client thread to flatten the map of futures into a single future and then send it back, i.e. we let the client thread to make one blocking call, so that the user thread, after getting the result, only need to make one more blocking call.

EDIT: thinking about it once again, now I think it could work after all: after we have collected the list of nodes, and hence we know how many futures we will have under the cover, we can do the flattening as I showed in my PR by:

1. have the single umbrella `future` registered all per-node future as its waiters.
2. have another `Applicant` future which register the umbrella `future` as its waiter; when the umbrella calls its `complete`, then this waiter will be triggered to fill in the collection via the given function.

So, if we call on the umbrella `future.get()` from the user thread, it will wait for all the per-node future to complete, fill in the result collection, and then return. So the flattening result in a single blocking call from user's point of view.

rajinisivaram committed:
KAFKA-4292: Configurable SASL callback handlers (KIP-86) (#2022)

Implementation of KIP-86. Client, server and login callback handlers have been made configurable for both brokers and clients.

Reviewers: Jun Rao <junrao@gmail.com>, Ron Dagostino <rndgstn@gmail.com>, Manikumar Reddy <manikumar.reddy@gmail.com>

cmccabe committed:
KAFKA-6694: The Trogdor Coordinator should support filtering task responses (#4741)

omkreddy committed:
KAFKA-6741:  Disable Selector's idle connection timeout in testNetworkThreadTimeRecorded() test (#4824)

Reviewers: Jason Gustafson <jason@confluent.io>, Rajini Sivaram <rajinisivaram@googlemail.com>

omkreddy committed:
MINOR: Update max.connections.per.ip.overrides config docs (#4819)

Add a validation check to make sure max.connections.per.ip.overrides is configured when max.connections.per.ip is set zero. Also clean up the config description.

cmccabe commented:
Check out `KafkaAdminClient#deleteRecords` for an example of a two-step process of finding partition leaders and then deleting records from them, asynchronously.

Commit:
KAFKA-6748: double check before scheduling a new task after the punctuate call (#4827)

After the punctuate() call, we would like to double check on the scheduled flag since the call itself may cancel it.

Reviewers: Guozhang Wang <wangguoz@gmail.com>, John Roesler <john@confluent.io>

guozhangwang commented:
Thanks @cmccabe. I think the pattern of `describeConsumer` could be similar to `deleteRecords`, but there is still a small difference between and `listConsumers` and the former two: for `listConsumers` the possible keys of the futures map, i.e. size of the flattened collection is unknown beforehand (note for `describeConsumer` we have the queried collection of groups from the input, so we should be able to build the future at the very beginning, and also we can group the requests per consumer coordinator like `deleteRecords` did; in that way, the map contained in the result could be `Map<String, KafkaFuture<ConsumerGroupDescription>>`), so we can only populate and flatten the future after the first round.

That being said, what I've suggested in jeqo#2 should not introduce a blocking call on the client thread (I was unclear about `addWaiters` behavior), all it does, is that after the first round trip, we have known how many consumer coordinators are there, construct an umbrella future that enwraps all the requested futures, and add another function via `thenApply` to populate the results once every covered future has completed.

Anyways, I'll leave it for @jeqo to explore for possible ways to flatten.

tedyu committed:
KAFKA-6747 Check whether there is in-flight transaction before aborting transaction (#4826)

As Frederic reported on mailing list under the subject "kafka-streams Invalid transition attempted from state READY to state ABORTING_TRANSACTION", producer#abortTransaction should only be called when transactionInFlight is true.

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Matthias J. Sax <matthias@confluent.io>

apovzner committed:
Trogdor: Added commonClientConf and adminClientConf to workload specs (#4757)

Currently, WorkerUtils will be able to create topics when there is no security. To be able to work with secure kafka, WorkerUtils.createTopic() needs to be able to take security configs. This PR adds commonClientConf field to both producer bench and roundtrip workload specs so that users can specify security and other common configs once for producer/consumer and adminClient. Also added adminClientConf field to workload specs so that users can specify adminClient specific configs if they want to. For completeness, added consumerConf and producerConf to roundtrip workload spec.

Reviewers: Ismael Juma <ismael@juma.me.uk>, Colin P. Mccabe <cmccabe@confluent.io>, Rajini Sivaram <rajinisivaram@googlemail.com>

rajinisivaram committed:
KAFKA-6576: Configurable Quota Management (KIP-257) (#4699)

Enable quota calculation to be customized using a configurable callback. See KIP-257 for details.

Reviewers: Jun Rao <junrao@gmail.com>

mjsax committed:
KAFKA-6054: Fix upgrade path from Kafka Streams v0.10.0 (#4779)

Reviewers: Guozhang Wang <guozhang@confluent.io>, Bill Bejeck <bill@confluent.io>, John Roesler <john@confluent.io>, Damian Guy <damian@confluent.io>

ijuma committed:
MINOR: Mention leader in a few follower/controller log messages (#4835)

cmccabe committed:
KAFKA-6688. The Trogdor coordinator should track task statuses (#4737)

Reviewers: Anna Povzner <anna@confluent.io>, Rajini Sivaram <rajinisivaram@googlemail.com>

asdf2014 committed:
MINOR: Remove magic number and extract Pattern instance from method as class field (#4799)

* Remove magic number
* Extract Pattern instance from method as class field
* Add @Override declare

Reviewers: Randall Hauch <rhauch@gmail.com>

ijuma committed:
MINOR: Java 10 fixes so that the build passes (#4839)

* Upgrade EasyMock to 3.6 which adds support for Java 10
by upgrading to ASM 6.1.1.

* Ensure that Jacoco is truly disabled for the `core` project.
This was the original intent, since it's in Scala, but it had not
been achieved. This is important because the Jacoco agent
fails when it tries to instrument the classes compiled by
scalac with Java 10.

cmccabe commented:
> Thanks @cmccabe. I think the pattern of describeConsumer could be similar to deleteRecords, but there is still a small difference between and listConsumers and the former two: for listConsumers the possible keys of the futures map, i.e. size of the flattened collection is unknown beforehand (note for describeConsumer we have the queried collection of groups from the input, so we should be able to build the future at the very beginning, and also we can group the requests per consumer coordinator like deleteRecords did; in that way, the map contained in the result could be Map<String, KafkaFuture<ConsumerGroupDescription>>), so we can only populate and flatten the future after the first round.

Hmm.  Why does there have to be a futures map for `listConsumerGroups` in the first place?  `listConsumerGroups` just returns a future which contains the list of all the consumer groups in the cluster, right?  I don't see why a map needs to be involved.  On the other hand, if you want to learn more about a specific consumer group, you use `describeConsumerGroup`, which does know the groups it is getting information about ahead of time.  This is analogous to `listTopics` versus `describeTopics`.  `listTopics` doesn't have a map.  `describeTopics` does.

> That being said, what I've suggested in jeqo#2 should not introduce a blocking call on the client thread (I was unclear about addWaiters behavior), all it does, is that after the first round trip, we have known how many consumer coordinators are there, construct an umbrella future that enwraps all the requested futures, and add another function via thenApply to populate the results once every covered future has completed.

Yes, it's very crucial to avoid blocking the admin client thread, or any function using the admin client.

@jeqo, I think if you look at the other admin client functions, you should be able to see how this can work.  Let me know if you need any help with this.  And thanks again for working on KIP-222.

guozhangwang commented:
> Hmm. Why does there have to be a futures map for listConsumerGroups in the first place? listConsumerGroups just returns a future which contains the list of all the consumer groups in the cluster, right? I don't see why a map needs to be involved. On the other hand, if you want to learn more about a specific consumer group, you use describeConsumerGroup, which does know the groups it is getting information about ahead of time. This is analogous to listTopics versus describeTopics. listTopics doesn't have a map. describeTopics does.

Just to clarify, we will not expose the map inside the future to user; the map will be maintained internally only during the second round-trip as we need to check if we have received the responses from all found broker nodes, but as I suggested we should flatten it before exposing to the users. So the returned type of `listConusmer` will be `KafkaFuture<Collection<ConsumerGroupListing>>`, while the returned type of `describeConsumer` will be `Map<String, KafkaFuture<ConsumerGroupDescription>>` (note for the latter, it is because users already specifically give the list of consumer group ids, like the `deleteRecords` api)

hachikuji committed:
KAFKA-6768; Transactional producer may hang in close with pending requests (#4842)

This patch fixes an edge case in producer shutdown which prevents `close()` from completing due to a pending request which will never be sent due to shutdown initiation. I have added a test case which reproduces the scenario.

Reviewers: Apurva Mehta <apurva@confluent.io>, Ismael Juma <ismael@juma.me.uk>

apovzner committed:
Kafka-6693: Added consumer workload to Trogdor (#4775)

Added consumer only workload to Trogdor. The topics must already be pre-populated. The spec lets the user request topic pattern and range of partitions to assign to [startPartition, endPartition].

Reviewers: Colin P. Mccabe <cmccabe@confluent.io>, Rajini Sivaram <rajinisivaram@googlemail.com>

edenhill committed:
Make [Config]Resource.toString() consistent with existing code (#4845)

The toString() for ConfigResource was using { } instead of ( ) which is inconsistent with the existing toStrings in the code, while toString for Resource was using a mix of ( and }.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupsResult.java:
```diff
@@ -0,0 +1,59 @@
+        this.futures = futures;
+    }
+
+    public KafkaFuture<Map<String, KafkaFuture<Void>>> values() {
```
I think we can just have one function between `values` and `groups` here. I'd suggest we use

```
public Map<TopicPartition, KafkaFuture<Void>> deletedGroups()
```

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeConsumerGroupsResult.java:
```diff
@@ -0,0 +1,69 @@
+        return futures;
+    }
+
+    public KafkaFuture<Collection<String>> names() {
```
Where is this function used?

I'd suggest we only keep one function, i.e.

```
public Map<TopicPartition, KafkaFuture< ConsumerGroupDescription >> DescribeConsumerGroupsResult#values()
```

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupsResult.java:
```diff
@@ -0,0 +1,59 @@
+        });
+    }
+
+    public KafkaFuture<Collection<KafkaFuture<Void>>> all() {
```
For `all()` function, its returned type should be `KafkaFuture<Void>`; ditto for other two Results as well.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupOffsetsResult.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Return a future which yields a map of topic partitions to OffsetAndMetadata objects.
+     */
+    public KafkaFuture<Map<TopicPartition, OffsetAndMetadata>> partitionsToOffsetAndMetadata() {
```
I'd suggest only keep `partitionsToOffsetAndMetadata` here.

guozhangwang reviewed (commented):
@jeqo I left some more comments on the return types of other Result classes.

Also the jenkins failures seem consistent, need to fix them before merging as well.

rajinisivaram committed:
MINOR: Move creation of quota callback to ensure single instance (#4848)

Move creation of quota callback instance out of KafkaConfig constructor to QuotaFactory.instantiate to avoid creating a callback instance for every KafkaConfig since we create temporary KafkaConfigs during dynamic config updates.

Reviewers: Jun Rao <junrao@gmail.com>

omkreddy committed:
KAFKA-4883: handle NullPointerException while parsing login modue control flag (#4849)

cmccabe committed:
MINOR: Add NullPayloadGenerator to Trogdor (#4844)

jeqo committed:
fix test with updated method firm

jeqo committed:
remove redundant access methods

jeqo committed:
remove redundant access methods

jeqo committed:
remove unused imports

jeqo committed:
ignoring test to be fixed on follow-up PR

guozhangwang commented:
retest this please

omkreddy committed:
KAFKA-6752: Enable unclean leader election metric (#4838)

Reviewers: Jun Rao <junrao@gmail.com>

omkreddy committed:
KAFKA-6447: Add Delegation Token Operations to KafkaAdminClient (KIP-249) (#4427)

Reviewers: Jun Rao <junrao@gmail.com>

guozhangwang commented:
@jeqo could you rebase your PR?

guozhangwang committed:
rebase from trunk

guozhangwang commented:
> @jeqo could you rebase your PR?

NVM, I've done the rebase, and will merge i once the tests passed locally.

guozhangwang commented:
Unit test passed locally, merging now

jeqo commented:
@guozhangwang thanks!! :)
